sled-agent/config-reconciler/src/reconciler_task/zones.rs

```diff
@@ -0,0 +1,1119 @@
+    use illumos_utils::link::VnicAllocator;
+    use illumos_utils::running_zone::ZoneBuilderFactory;
+    use illumos_utils::zpool::PathInPool;
+    use illumos_utils::zpool::ZpoolName;
+    use illumos_utils::zpool::ZpoolOrRamdisk;
+    use nexus_sled_agent_shared::inventory::OmicronZoneDataset;
+    use nexus_sled_agent_shared::inventory::OmicronZoneImageSource;
+    use omicron_common::address::SLED_PREFIX;
+    use omicron_test_utils::dev;
+    use omicron_uuid_kinds::ZpoolUuid;
+    use std::collections::BTreeSet;
+    use std::collections::VecDeque;
+    use std::sync::Mutex;
+
+    // Helper to construct a `RunningZone` even on non-illumos systems.
+    struct FakeZoneBuilder {
+        vnic_alloc: VnicAllocator<Etherstub>,
+        factory: ZoneBuilderFactory,
+        tempdir: Utf8TempDir,
+    }
+
+    impl FakeZoneBuilder {
+        fn new() -> Self {
+            let vnic_source = Etherstub("teststubvnic".to_string());
+            let vnic_alloc = VnicAllocator::new(
+                "testvnic",
+                vnic_source,
+                illumos_utils::fakes::dladm::Dladm::new(),
+            );
+            let tempdir =
+                Utf8TempDir::with_prefix("test-config-reconciler-zones-")
+                    .expect("created temp dir");
+            let factory = ZoneBuilderFactory::fake(
+                Some(tempdir.path().as_str()),
+                illumos_utils::fakes::zone::Zones::new(),
+            );
+            Self { vnic_alloc, factory, tempdir }
+        }
+
+        async fn make_running_zone(
+            &self,
+            name: &str,
+            log: Logger,
+        ) -> RunningZone {
+            let installed_fake_zone = self
+                .factory
+                .builder()
+                .with_zone_type(name)
+                .with_zone_root_path(PathInPool {
+                    pool: ZpoolOrRamdisk::Ramdisk,
+                    path: "/test-zone-root".into(),
+                })
+                .with_underlay_vnic_allocator(&self.vnic_alloc)
+                .with_log(log)
+                .with_opte_ports(Vec::new())
+                .with_links(Vec::new())
+                .install()
+                .await
+                .expect("installed fake zone");
+            RunningZone::fake_boot(0, installed_fake_zone)
+        }
+    }
+
+    #[derive(Debug, Default)]
+    struct FakeZoneFacilities {
+        inner: Mutex<FakeZoneFacilitiesInner>,
+    }
+
+    #[derive(Debug, Default)]
+    struct FakeZoneFacilitiesInner {
+        halt_responses: Option<VecDeque<Result<(), ZoneShutdownError>>>,
+        removed_gz_addresses: BTreeSet<AddrObject>,
+    }
+
+    impl FakeZoneFacilities {
+        fn push_halt_response(&self, response: Result<(), ZoneShutdownError>) {
+            let mut inner = self.inner.lock().unwrap();
+            inner.halt_responses.get_or_insert_default().push_back(response);
+        }
+    }
+
+    impl ZoneFacilities for FakeZoneFacilities {
```

Comment:
This trait based fake impl setup is pretty nice compared to what we had before.

The tests are really easy to read and reason about. In fact I read them before the code and that worked well.